### DISCUSSION TITLE
Flux and Helm

### DISCUSSION THREAD

**Discussion by tcheronneau** [fluxcd/flux2#2125] [Q&A] [answered]
Hey there, 

We are using Flux since a moment now and it was working quite great with kustomize so far. 
We've decided to move to Helm and the migration is really not as smooth as expected compared to what we were getting with Kustomize. 

Our Helm controller is 0.12.1 and kubernetes version is 1.20.11. 

We had several issues : 
- If we delete a deployment which is part of a helm release it's not put back by Flux (only if you modify the helmrelease file). 
- I just hit something quite weird today when deleted all pod from a namespace deployed by a helm release : it deleted also service account and configmaps. 
- If I override a version of a chart which is deployed by Flux it's not detected and so not redeployed. 

I think it's all the issues / complains we had so far if we compare to kustomize (maybe helm is requiring more configuration that I did not notice so sorry if it's the case). 

Thanks,
Thomas.

**Comment by hiddeco** [comment, marked as answer]
> - If we delete a deployment which is part of a helm release it's not put back by Flux (only if you modify the helmrelease file).

This is a known limitation, see: https://github.com/fluxcd/helm-controller/issues/186#issuecomment-932107655

> - I just hit something quite weird today when deleted all pod from a namespace deployed by a helm release : it deleted also service account and configmaps. 

Without providing precise instructions about the actions you took, and how they are related to a chart, it is hard to say anything useful about this. However, you generally should not manage namespaces through charts.

> - If I override a version of a chart which is deployed by Flux it's not detected and so not redeployed.

What version did you overwrite? As this can refer to multiple things. Generally, the helm-controller should perform an upgrade if you perform a manual `helm upgrade` (which overwrites the version).

**Reply by kingdonb** [reply]
> I just hit something quite weird today when deleted all pod from a namespace deployed by a helm release : it deleted also service account and configmaps.

As for this issue, it's the only one I can't explain from your original post @tcheronneau. I think your question about changes not being deployed is the same one I answered just now, over here: https://github.com/fluxcd/flux2/discussions/2242

I'm going to mark this question as answered, but if you have more questions feel free to ask!

**Comment by stefanprodan** [comment]
> We've decided to move to Helm and the migration is really not as smooth as expected compared to what we were getting with Kustomize.

You can't fully replace kustomize overlays with Helm charts. Many things do no belong in a Helm chart, such as namespaces, CRDs and other cluster-wide resources. You could consider using Helm to package the app deployments which should be made out of namespace-scoped resources.

**Comment by pkit** [comment]
I think using helm is a kind of a dead end now for GitOps flow.
I'm just converting all the "helm-only" charts to `kustomize` manifests and use it that way. You also get a better visibility on what exactly changed between helm releases.
Unfortunately there are no tools to do it automatically (`helm convert` is outdated and unmaintained) but it's pretty easy to write scripts that will do it. Will contribute it later if needed.

**Reply by gcaracuel** [reply]
> I'm just converting all the "helm-only" charts to `kustomize` manifests and use it that way. You also get a better visibility on what exactly changed between helm releases.

https://github.com/databus23/helm-diff can help with this, both `helm diff release` or `helm diff revision` could match depending on your use case.

**Reply by pkit** [reply]
Thanks, but I'm not sure how to use it?
For example: how can I get what will be installed if I don't use helm releases? And install everything through kustomize.

**Reply by kingdonb** [reply]
Kustomize and Helm are not directly substitutable for one another – Helm has features which cannot be emulated in Kustomize, like post-install hooks, and vice-versa (in this comparison of the two tools, the transparency that you're seeking itself is actually considered as an important feature of Kustomize compared to Helm: https://stackoverflow.com/a/70423452/661659)

This is why Helm Controller is based on the Helm package – so it can perfectly emulate Helm's actual behavior, (it's not even fair to call it emulation – Helm is doing the work.) It is a double-edged sword though. If there is a problem with Helm, it will surface in Helm Controller just the same. Helm releases are difficult to inspect, both before and after they process.

You could always run `helm template` in a CI job and use a workflow like this one to commit them back to Git:

https://github.com/kingdonb/scrob-web/blob/95f625f1022e47029d4bc35c8d37d241f3d37569/.github/workflows/kuby.yml#L87-L93

(but it won't be exactly what's installed by Helm, unless your charts are not using any of the more "interesting" features of Helm.)

**Reply by pkit** [reply]
Yup. I agree. The problem here is that in the GitOps paradigm all the "interesting" features of Helm are a thing to avoid, because it removes the declarative aspect. But inability to even reconcile a cluster when something is wrong/missing/changed in Helm is a non-starter for the `helm-controller`.

**Reply by vterdunov** [reply]
Agreed, imho current helm implementation doesn't fit to Gitops paradigm. Im trying to switch to helm template+kustomize.
For me, most interesting and useful thing from helm is hooks. It allows safely and once to run DB migrations. However, I don't know how to properly achieve that using kustomize. As far as I investigated there is problem to run Job+init container with k8s-wait script.

**Reply by kingdonb** [reply]
Database migrations are not exactly declarative according to this logic, but I would not tell anyone to avoid using database migrations for this reason. The interesting features remain important nonetheless; although I don't use these features of Helm either; I prefer to get my database migrations done with an init container now!

In any case, the behavior is an aspect of Helm, and so I suspect you would need to fork Helm in order to change that. It should be clear, the position of Flux is that Helm Controller doesn't fork Helm.

This method works with helm or kustomize, or neither if you prefer; it's for Rails, and `kuby` is for Ruby, but it requires that your migrate is idempotent and safe to run at the start of every web server process

https://github.com/kingdonb/scrob-web/blob/95f625f1022e47029d4bc35c8d37d241f3d37569/manifests/k8s.yml#L95-L117

Is this an approach you think could work?

**Reply by vterdunov** [reply]
@kingdonb thanks for reply! There are couple of problems with this "naked" init containers solution.
Init containers will be run on any pod creation event. Including deployment update and replicas auto-scale.
Unfortunately, not all db migration systems support multiple runs at the same time. In other words, they cannot set sort of lock in database.
